### ISSUE TITLE
ReceiveAsync issues

### ISSUE THREAD

**Issue by zbynek001** [akkadotnet/akka.net#3259]
we've run into several issues with current ReceiveAsync implementation:
- It's breaking the principle of processing one message at a time
  - During processing of the message in ReceiveAsync with several continuations, the actor is processing system messages in between task continuations, which can leads to the second issue

- Under some condition, the scheduled tasks are not executed at all
  - This one is more severe and we've run into it several time in production. Because the actor is processing system messages in between task continuations, actor can be terminated before finishing processing on ReceiveAsync. This leads to sending the task continuation to DeadLetter. Which means the task will never get executed. And if this task is some global initialization task, this will screw the whole application, because other parts will be waiting for it but i'll never finish.

The way how I've fixed this now is:

- Adding another suspend state ("stop") to maibox which also stops processing system messages
- ActorTaskScheduler is stopping the mailbox, including system messages
- ActorTaskScheduler is not scheduling continuations via system messages but directly on ThreadPool
- ActorTaskScheduler concurrency is limited to 1

So far it works fine and also all tests are passing.
Would you be interested in MR for this or you see some potential issues handling it this way?

**Comment by stijnherreman** [comment]
For your first point: if you want it to block, use `await` instead of continuations.

**Comment by zbynek001** [comment]
of course I'm using await. But this issue is related to how ActorTaskScheduler is scheduling task continuations, including await.

**Comment by stijnherreman** [comment]
Can you show an example of the code that doesn't work as desired? I don't experience such problems when using await.

**Comment by zbynek001** [comment]
no, don't have any sample for this and normally you won't run into this issue and it's very difficult to reproduce under normal conditions. Theoretically you could reproduce it with await delay and calling Stop in between.
In our case, we were running into it together with CosmosDb sdk, which is using quite many task continuations. And it didn't happen very often, like once a week, but when it happened, there was no way back except restarting the app.

**Comment by nvivo** [comment]
@zbynek001 

> During processing of the message in ReceiveAsync with several continuations, the actor is processing system messages in between task continuations, which can leads to the second issue

I guess this is expected due to how akka handles async messages, as it breaks continuations into messages to itself. The real solution to this would be to simply let the dispatcher to schedule tasks instead of actions, and propagate tasks all the way up. But this would break compatibility with so many stuff that it would have to be Akka 2.0. This was discussed a few times.

In any case, I have lots of actors running initialization with async code, and never had any real problems. My solution is to never terminate these type of actors abruptly, instead send a message telling it to shutdown (not a system message, a normal one) and let it handle with a `Context.Stop(Self)` at the right moment.

**Comment by Aaronontheweb** [comment]
> During processing of the message in ReceiveAsync with several continuations, the actor is processing system messages in between task continuations, which can leads to the second issue

The original `ReceiveAsync` implementation did not do this, but when I reimplemented the `ActorCell` semantics as part of Akka.NET v1.1 I made a choice to allow `SystemMessage`s to still be processed ahead of async continuations. Reasons for this boiled down to allowing async actors who are waiting on a continuation that is hopeless (i.e. didn't set a timeout and the request will never complete) to still be terminated if necessary. 

>In any case, I have lots of actors running initialization with async code, and never had any real problems. My solution is to never terminate these type of actors abruptly, instead send a message telling it to shutdown (not a system message, a normal one) and let it handle with a Context.Stop(Self) at the right moment.

I do the same thing; I have many actors that work with I/O drivers and `WebSocket` instances, and when I use `ReceiveAsync` I always use a `PoisonPill` instead of `Context.Stop` for this reason, since the `PoisonPill` goes through the user messages queue.

**Comment by zbynek001** [comment]
ahh, ok, it's a feature, not a bug :)
If there would be at least some way to mark Task as cancelled when the actor was terminated, but that's not possible.

Anyway, in our case, I'm not calling Stop in the middle of the async operation and still sometimes the `ActorTaskSchedulerMessage` ends up in deadletters. And the only possibility is, that the actor was terminated before `ReceiveAsync` is finished. I've spend quite a lot of time on this and didn't find any better way to handle it than what I mentioned above, so for now, will keep using it.

**Comment by nvivo** [comment]
@zbynek001 It's not normal for actors to be terminated at random, something must be telling them to do so. If this really happens, it would be good if you could reproduce it.

Could it be that the parent is stopping and being restarted?

**Comment by zbynek001** [comment]
hmm, when you mentioned the parent, I might have an idea.

In our case, the parent is sharded actor.
Could this be happening? :

1. Parent receives message
1. Parent creates child actor and forwards it the message
1. The child actor is processing the message via ReceiveAsync
1. Before the processing is finished, the sharding will rebalance a region with the parent actor
1. This will terminate the child actor before ReceiveAsync is finished

**Comment by Horusiath** [comment]
@zbynek001 in your case this could have happened if you're using default stop message (which is a `PoisonPill`) - in this case upon rebalancing, a parent actor will simply be killed, along with its children.

You could set up a custom stopping message:
1. When parent (sharded entity) receives that stopping message, it sends some kind of a token message to all of its children and change it's behavior to stopping.
2. When child is processing something asynchronously, it won't consume any user messages for this time. But when it'll finish, it can pick up the next message (at some point it will be a token). If a token will be received, it can i.e. stop itself.
3. When parent detects a `Terminated` message while in stopping behavior, it checks if all children have been stopped. If they had, it will stop itself and shard will be able to be correctly rebalanced.

Example:

```csharp
// shard initialization
var region = await ClusterSharding.Get(system).StartAsync(
    typeName: nameof(Parent),
    entityProps: Props.Create<Parent>(),
    settings: ClusterShardingSettings.Create(system),
    messageExtractor: new MessageExtractor(),
    allocationStrategy: new LeastShardAllocationStrategy(),
    handOffMessage: CustomStop.Instance);
// parent
class Parent : ReceiveActor
{
    public Parent()
    {
       Receive<CustomStop>(_ =>{
           foreach (var child in Context.Children()) child.Tell(Token.Instance);
           Become(Stopping);
       });
    }
    void Stopping() 
    {
        Receive<Terminated>(_ => { 
            if (Context.Children().Count == 0) Context.Stop(Self);
        });
    }
}
// child 
class Child: ReceiveActor
{
    public Child()
    {
        ReceiveAsync<MyMessage>(async msg => {});
        ReceiveAsync<Token>(_ => Context.Stop(Self));
    }
}
```

**Comment by zbynek001** [comment]
Ok, great. If this can really happen than it might explain our issue.

We're using custom message, so we could coordinate the stop our self.
But then this will create another issue in case the ReceiveAsing takes too long. Because the sharding will not send any new message to the actor until the actor is terminated, if I remember correctly, but that's a different issue. If there would be a way to cancel the ongoing task...
Will have to think what would be the best approach to take for us here.

Anyway, thanks a lot

**Comment by nvivo** [comment]
@zbynek001 you could set a `CancellationTokenSource` in your coordinator or even outside of akka, and pass the token to these tasks. 

I use this strategy to get out of some loops and delays when the process is about to stop.
